src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs

```diff
@@ -2528,5 +2528,19 @@ internal override Symbol RemapSymbolIfNecessaryCore(Symbol symbol)
 
             return null;
         }
+
+        internal override bool ShouldSkipSyntaxNodeAnalysis(SyntaxNode declaredNode, ISymbol declaredSymbol)
+        {
+            // Skip the topmost record declaration syntax node when analyzing synthesized record declaration constructor
+            // to avoid duplicate syntax node callbacks.
+            // We will analyze this node when analyzing the record declaration type symbol.
+            if (declaredNode.Kind() is SyntaxKind.RecordDeclaration or SyntaxKind.RecordStructDeclaration &&
+                declaredSymbol.GetSymbol() is SynthesizedRecordConstructor)
```

Comment:
It feels like we can simplify and align this check with how the helper above is doing that. This will make it easier to see relationship between the two and keep them aligned in the future. I am thinking about something like ```declaredNode is RecordDeclarationSyntax && declaredSymbol.Kind is SymbolKind.Method```